service/history/mutableStateBuilder.go

```diff
@@ -1280,6 +1359,65 @@ func (e *mutableStateBuilder) AddRequestCancelExternalWorkflowExecutionFailedEve
 	return nil
 }
 
+func (e *mutableStateBuilder) AddSignalExternalWorkflowExecutionInitiatedEvent(decisionCompletedEventID int64,
+	signalRequestID string, request *workflow.SignalExternalWorkflowExecutionDecisionAttributes) *workflow.HistoryEvent {
+
+	event := e.hBuilder.AddSignalExternalWorkflowExecutionInitiatedEvent(decisionCompletedEventID, request)
+	if event == nil {
+		return nil
+	}
+
+	initiatedEventID := *event.EventId
+	ri := &persistence.SignalInfo{
+		InitiatedID:     initiatedEventID,
+		SignalRequestID: signalRequestID,
+		SignalName:      request.GetSignalName(),
+		Input:           request.Input,
```

Comment:
Probably you need to store Control as mutable state too so it can be used for Delivered or Failed event.

Reply:
Added